### ISSUE TITLE
missing field `unused` for key `patch`

### ISSUE THREAD

**Issue by jmgilman** [nix-community/naersk#308]
```
 > error: failed to parse lock file at: /build/dummy-src/Cargo.lock
 >
 > Caused by:
 >   missing field `unused` for key `patch`
 > [naersk] cargo returned with exit code 101, exiting
```

I am getting the above error when building a lockfile with entries like:

```
[[patch.unused]]
name = "package"
version = "0.1.0"
source = "git+https://github.com/org/repo.git?branch=mybranch#efe089af522e424d1099407b6585efe042ad085f"
```

I can run this same lockfile with cargo outside of Nix and it compiles fine. I'm having trouble understanding what specifically is making it break when trying to build it with `naersk`.

**Comment by Patryk27** [comment]
Curiously, there's one-and-a-half worth of bugs here!

tl;dr [I've just prepared a fix](https://github.com/nix-community/naersk/pull/309) - after it's merged, you'll have to bump your Nixpkgs to a newer version (at least 23, i.e. current `master` or `nixpkgs-unstable`), and then bump Naersk.

---

First, when Naersk builds a derivation, it converts `Cargo.lock` into a Nix attrset and then reserializes that attrset back into a TOML document later - this serialization-code has had a bug that caused the eventual `Cargo.lock` to look like:

```toml
[patch]
[[unused]]
name = "package"
version = "0.1.0"
source = "..."
```

... which Cargo (rightfully) doesn't understand anymore.

This bug has been [fixed recently](https://github.com/nix-community/naersk/pull/300), but the fix only works when Naersk is launched on Nixpkgs 23+ (because older versions of Nixpkgs had a buggy serializer as well) - so, in principle, if you switched to newer Nixpkgs (e.g. `nix flake lock --update-input naersk --update-input nixpkgs`), it should work.

But, as it turns out, there's second thing at play here - because Naersk doesn't provide sources for ignored dependencies, trying to actually compile such project will fail, saying:

```
failed to resolve address for github.com: nodename nor servname provided, or not known; class=Net (12)
```

... as Cargo will try to download the "missing" dependency on its own, while being locked in Nix sandbox, which fails.

This is a bug on Naersk's side, for which I've just prepared a fix.
